src/app/shared/new-version/new-version.component.ts

```diff
@@ -16,185 +16,191 @@ limitations under the License.
-        label: this.form.label,
-        copyPermissions: this.form.copyPermissions,
-        copyDataFlows: this.form.copyDataFlows
+
+    this.processing = true;
+    this.setupForm.disable();
+
+    if (this.typeValue === 'Fork') {
+      const resource: ForkModelPayload = {
+        label: this.label.value,
+        copyPermissions: false,
+        copyDataFlows: false
       };
-      this.processing = true;
-
-      this.resources[this.domainType]
-        .newForkModel(this.catalogueItem.id, resource)
-        .pipe(finalize(() => (this.processing = false)))
-        .subscribe(
-          (response) => {
-            this.messageHandler.showSuccess(
-              'New Data Model version created successfully.'
-            );
-            this.stateHandler.Go(
-              this.domainType,
-              { id: response.body.id },
-              { reload: true }
-            );
-          },
-          (error) => {
-            this.processing = false;
-            this.messageHandler.showError(
-              'There was a problem creating the new Data Model version.',
-              error
-            );
-          }
-        );
-    } else if (this.versionType === 'Version') {
-      this.processing = true;
-      this.resources[this.domainType]
-        .newBranchModelVersion(this.catalogueItem.id, {})
-        .subscribe(
-          (response) => {
-            this.processing = false;
-            this.messageHandler.showSuccess(
-              'New Document Model version created successfully.'
-            );
-            this.stateHandler.Go(
-              this.domainType,
-              { id: response.body.id },
-              { reload: true }
-            );
-          },
-          (error) => {
-            this.processing = false;
-            this.messageHandler.showError(
-              'There was a problem creating the new Document Model version.',
-              error
-            );
-          }
-        );
-    } else if (this.versionType === 'Branch') {
-      let resources = {};
-      if (this.form.label !== null && this.form.label !== '') {
-        resources = { branchName: this.form.label };
-      }
 
-      this.processing = true;
-      this.resources[this.domainType]
-        .newBranchModelVersion(this.catalogueItem.id, resources)
-        .subscribe(
-          (response) => {
-            this.processing = false;
-            this.messageHandler.showSuccess('New Branch created successfully.');
-            this.stateHandler.Go(
-              this.domainType,
-              { id: response.body.id },
-              { reload: true }
-            );
-          },
-          (error) => {
-            this.processing = false;
-            this.messageHandler.showError(
-              'There was a problem creating the new Document Model version.',
-              error
-            );
-          }
-        );
+      const request = this.resources[this.domainElementType.resourceName]
```

Comment:
Switch Case would be cleaner here

Reply:
After this comment I tried building up some adapter services around the `MdmResourcesService` which would expose common endpoint functions e.g. `newBranchModelVersion()` and then use `switch-case` statements to directly drill down to the correct resource based on domain type.

After trying this though it felt like a lot of extra code to write when the original way of property indexing was shorter and faster. Bear in mind that every endpoint with a similar signature has to have their own `switch-case` block. I'm also a bit concerned about how this would minimise during a production build, it could bloat the final distributable output.

I would suggest leaving this for now, but maybe add a new issue under the "Clean up" label/project to discuss later.

Reply:
agreed